### PR TITLE
Update 1D Profile Viewer to use wcsaxes for plotting 1D spectrum of data cubes

### PR DESCRIPTION
**Pull request by kakirastern**
## Description

Tasks accomplished in this PR:
1. Update the existing `glue`'s "1D Profile" viewer to use Astropy's `wcsaxes` for plotting. 
2. Add a `slice` function option to the "1D Profile" to enable plotting a 1D spectrum of a data cube (e.g. a raster data cube) at a position picked out by `glue-solar`'s pixel extraction tool. 
3. Enable deletion of `IndexedData` type data so that now the dataset in the data collection generated by `glue-solar`'s pixel extraction tool can be deleted.

### CODE REVIEW

**Comment by kakirastern** [comment]
Am wondering if it's about time to upstream `glue-solar`'s pixel extraction tool to here, since will need this tool for the missing `indices` variable that is causing the CI tests to fail.

**Comment by kakirastern** [comment]
@astrofrog @Cadair If I can get some guidance as to how to approach the debugging of this PR to comply with `glueviz` standards so that it can be merged eventually it would be great. I have tested it and everything seems to be in working order.

**Comment by kakirastern** [comment]
Yay, the Travis CI tests are passed finally!

File: glue/app/qt/layer_tree_widget.py [hunk 1]
```diff
@@ -197,14 +198,14 @@ class DeleteAction(LayerAction):
 
     def _can_trigger(self):
         selection = self.selected_layers()
-        return all(isinstance(s, (core.Data, core.SubsetGroup))
+        return all(isinstance(s, (core.Data, core.SubsetGroup, IndexedData, SlicedData))
```
Comment (astrofrog) [inline glue/app/qt/layer_tree_widget.py:201]: I think here you should just replace Data, IndexedData and SlicedData by BaseData

File: glue/app/qt/layer_tree_widget.py [hunk 2]
```diff
@@ -197,14 +198,14 @@ class DeleteAction(LayerAction):
...
                    for s in selection)
 
     def _do_action(self):
         assert self._can_trigger()
         selection = self.selected_layers()
         for s in selection:
-            if isinstance(s, core.Data):
+            if isinstance(s, (core.Data, IndexedData, SlicedData)):
```
Comment (astrofrog) [inline glue/app/qt/layer_tree_widget.py:208]: As above

File: glue/core/data.py [hunk 3]
```diff
@@ -1761,6 +1763,9 @@ def compute_statistic(self, statistic, cid, subset_state=None, axis=None,
             data = self.get_data(cid, view=view)
             mask = None
 
+        # if statistic == 'slice':
+        #     data = self.get_data(cid, view=view)
+
```
Comment (astrofrog) [inline glue/core/data.py:1768]: Remove?

File: glue/core/data.py [hunk 4]
```diff
@@ -1788,6 +1793,7 @@ def compute_statistic(self, statistic, cid, subset_state=None, axis=None,
                                    positive=positive, percentile=percentile)
 
         if subarray_slices is None or axis is None:
+
```
Comment (astrofrog) [inline glue/core/data.py:1796]: Maybe revert this

File: glue/core/data.py [hunk 5]
```diff
@@ -1801,6 +1807,7 @@ def compute_statistic(self, statistic, cid, subset_state=None, axis=None,
             full_result = np.zeros(full_shape) * np.nan
             result_slices = [subarray_slices[idim] for idim in range(self.ndim) if idim not in axis]
             full_result[result_slices] = result
+
```
Comment (astrofrog) [inline glue/core/data.py:1810]: Revert

File: glue/core/data_derived.py [hunk 6]
```diff
@@ -194,3 +194,181 @@ def compute_histogram(self, *args, **kwargs):
         else:
             kwargs['subset_state'] &= self._indices_subset_state
         return self._original_data.compute_histogram(*args, **kwargs)
+
+
+class SlicedData(BaseCartesianData, HubListener):
```
Comment (astrofrog) [inline glue/core/data_derived.py:199]: I wonder if we might want to have a single class than handles both slicing and indexing 🤔  But ignore for now 😄

File: glue/core/data_derived.py [hunk 7]
```diff
@@ -194,3 +194,181 @@ def compute_histogram(self, *args, **kwargs):
...
+            msg = NumericalDataChangedMessage(self)
+            self.hub.broadcast(msg)
+
+    @property
+    def label(self):
+        slice = '(' + ','.join([':' if x == 0 else str(x) for x in self._slices]) + ')'
+        return self._original_data.label + slice
+
+    @property
+    def shape(self):
```
Comment (astrofrog) [inline glue/core/data_derived.py:294]: For now, this and several other properties below will just return the same as the original data, right?

File: glue/utils/array.py [hunk 8]
```diff
@@ -450,6 +451,9 @@ def compute_statistic(statistic, data, mask=None, axis=None, finite=True,
     # NOTE: this function should not ever have to use glue-specific objects.
     # The aim is to eventually use a fast C implementation of this function.
 
+    if statistic == 'slice':
+        return np.array(data, dtype=float)[axis]
```
Comment (astrofrog) [inline glue/utils/array.py:455]: Is this needed?

File: glue/utils/array.py [hunk 9]
```diff
@@ -474,11 +478,11 @@ def compute_statistic(statistic, data, mask=None, axis=None, finite=True,
             data = np.array(data, dtype=float)
             data[~keep] = np.nan
 
-        function = NAN_FUNCTIONS[statistic]
+        function = NAN_FUNCTIONS[statistic] or None
```
Comment (astrofrog) [inline glue/utils/array.py:481]: Just to check, why is this needed?

File: glue/viewers/image/qt/slice_widget.py [hunk 10]
```diff
@@ -92,7 +107,8 @@ def sync_sliders_from_state(self, *args):
                     world = world_axis(self.data.coords, self.data,
                                        pixel_axis=self.data.ndim - 1 - i,
                                        world_axis=self.data.ndim - 1 - i)
-                    world_unit = self.data.coords.world_axis_units[i]
+                    world_unit = self.data.coords.world_axis_units[self.data.ndim - 1 - i]
```
Comment (astrofrog) [inline glue/viewers/image/qt/slice_widget.py:110]: This might be fixed in master (by @jsub1) so will need rebasing

File: glue/viewers/image/qt/slice_widget.py [hunk 11]
```diff
@@ -142,7 +159,7 @@ class FakeViewerState(State):
     viewer_state.y_att = data.pixel_component_ids[3]
     viewer_state.slices = [0] * 5
 
-    widget = MultiSliceWidgetHelper(viewer_state)
+    widget = ImageMultiSliceWidgetHelper(viewer_state)
```
Comment (astrofrog) [inline glue/viewers/image/qt/slice_widget.py:162]: If possible it would be good to avoid the change in the class name unless absolutely needed - since this is public API that may be used by other plugins.

File: glue/viewers/profile/layer_artist.py [hunk 12]
```diff
@@ -147,16 +148,17 @@ def _update_visual_attributes(self):
 
         self.redraw()
 
-    def _update_profile(self, force=False, **kwargs):
+    def _update_profile(self, force=True, **kwargs):
```
Comment (astrofrog) [inline glue/viewers/profile/layer_artist.py:151]: Is this intentional? It is a big change.

File: glue/viewers/profile/qt/options_widget.py [hunk 13]
```diff
@@ -33,10 +36,24 @@ def __init__(self, viewer_state, session, parent=None):
 
         self.session = session
 
+        for dataset in self.session.data_collection:
+            if isinstance(dataset, IndexedData):
+                self.viewer_state.indices = dataset.indices
```
Comment (astrofrog) [inline glue/viewers/profile/qt/options_widget.py:41]: So is the idea that for indexdata for example, the sliders can be used in the image viewer to change the actual indices of the IndexedData?

File: glue/viewers/profile/qt/slice_widget.py [hunk 14]
```diff
@@ -0,0 +1,136 @@
...
+from glue.core.coordinate_helpers import dependent_axes, world_axis
+from glue.viewers.common.qt.data_slice_widget import SliceWidget
+from glue.viewers.image.state import AggregateSlice
+from glue.utils.decorators import avoid_circular
+from glue.core.data_derived import IndexedData, SlicedData
+
+__all__ = ['ProfileMultiSliceWidgetHelper']
+
+
+class ProfileMultiSliceWidgetHelper(object):
```
Comment (astrofrog) [inline glue/viewers/profile/qt/slice_widget.py:11]: Hard to see here, but what are the main differences compared to the image multi-slice widget helper class? Can they share a common base class to reduce duplication?

File: glue/viewers/profile/layer_artist.py (hunk 12 above)
Comment (kakirastern) [reply glue/viewers/profile/layer_artist.py:151]: Yeah, it is intentional. We would like the sliders to be able to update the 1D profiles automatically. This seems to be the most natural place/way to do so.

File: glue/viewers/image/qt/slice_widget.py (hunk 10 above)
Comment (kakirastern) [reply glue/viewers/image/qt/slice_widget.py:110]: Rebased with the changes in master picked up

Comment (kakirastern) [reply glue/viewers/image/qt/slice_widget.py:110]: Wait, think there was some problem with my rebase... let me fix it 😉

Comment (kakirastern) [reply glue/viewers/image/qt/slice_widget.py:110]: Confirmed changes picked up...

File: glue/viewers/image/qt/slice_widget.py (hunk 11 above)
Comment (kakirastern) [reply glue/viewers/image/qt/slice_widget.py:162]: Yes, agreed. Let me revert this.

File: glue/viewers/profile/qt/options_widget.py (hunk 13 above)
Comment (kakirastern) [reply glue/viewers/profile/qt/options_widget.py:41]: Yes, and `SlicedData` is to supplement `IndexedData` for handling the slicing of data cubes. So in terms of functionality these two are distinct.

File: glue/core/data_derived.py (hunk 6 above)
Comment (kakirastern) [reply glue/core/data_derived.py:199]: Okay, sure 👌🏼

File: glue/viewers/profile/qt/slice_widget.py (hunk 14 above)
Comment (kakirastern) [reply glue/viewers/profile/qt/slice_widget.py:11]: I just compared the `glue/viewers/profile/qt/slice_widget.py` to `glue/viewers/image/qt/slice_widget.py` and found there are 19 differences. In essence they are not really structured exactly the same, although they look very similar. For example, the `image` one has the `y_att` which the `profile` one does not have. Also some code snippet that works for the former does not work for the latter, and needs to be tweaked to make things work.

File: glue/core/data.py (hunk 3 above)
Comment (kakirastern) [reply glue/core/data.py:1768]: Yup

File: glue/core/data.py (hunk 4 above)
Comment (kakirastern) [reply glue/core/data.py]: Sure

File: glue/utils/array.py (hunk 8 above)
Comment (kakirastern) [reply glue/utils/array.py:454]: Yes

File: glue/utils/array.py (hunk 9 above)
Comment (kakirastern) [reply glue/utils/array.py]: Just reverted

File: glue/viewers/profile/qt/slice_widget.py (hunk 14 above)
Comment (astrofrog) [reply glue/viewers/profile/qt/slice_widget.py:11]: My reason for asking is because I want to also add sliders to the 3-d viewer :) I'll try and take a look at the differences later today (but fine to leave separate for now)

Comment (firmstudio-kakirastern) [reply glue/viewers/profile/qt/slice_widget.py:11]: I see

Comment (kakirastern) [reply glue/viewers/profile/qt/slice_widget.py:11]: Sure, that would be really cool

File: glue/core/data_derived.py (hunk 7 above)
Comment (kakirastern) [reply glue/core/data_derived.py:294]: Yes, think so
